plugins/engines/ospray/ispc/render/SimulationRenderer.ispc

```diff
@@ -106,308 +106,253 @@ inline void composite(const varying vec4f& src, varying vec4f& dst,
-    //                (uint64)floor(point.z) * dimensions.x * dimensions.y);
-
-    //            const uint8 voxelValue = self->volumeData[index];
-
-    //            const float normalizedValue =
-    //                self->colorMapSize * (voxelValue - self->colorMapMinValue)
-    //                /
-    //                self->colorMapRange;
-    //            const vec4f colorMapColor = self->colorMap[normalizedValue];
-    //            shadowIntensity += colorMapColor.w;
-    //        }
-    //    }
-    return shadowIntensity;
+        // Intersection with volume
+        shadowIntensity +=
+            getVolumeShadowContribution(volume, self, lightRay, sample);
+    }
+    return shadowIntensity * self->shadows;
 }
 
-inline varying vec4f getVolumeContribution(
+inline vec4f getVolumeContribution(
     Volume* uniform volume, const uniform SimulationRenderer* uniform self,
     const varying Ray& ray, varying ScreenSample& sample)
 {
-    //    if (!self->colorMap)
-    //        return make_vec4f(0.f, 1.f, 0.f, 0.f);
-
     // Find volume intersections
     float t0, t1;
-    //    vec3f aabbMin = self->volumeOffset;
-    //    vec3f aabbMax =
-    //        make_vec3f(self->volumeOffset) +
-    //        make_vec3f(self->volumeDimensions) * self->volumeElementSpacing;
     intersectBox(ray, volume->boundingBox, t0, t1);
-    // return make_vec4f(0.f, 0.f, 0.f, 0.f);
 
     // Ray marching
     vec4f pathColor = make_vec4f(0.f);
-    const float epsilon = max(0.01f, (t1 - t0) / self->volumeSamplesPerRay);
-    const float random = getRandomValue(sample, self->randomNumber) * epsilon;
-    t0 -= random;
-    t1 -= random;
-    for (float t = t0; t < t1 && pathColor.w < 1.f; t += epsilon)
+    const float epsilon = 1.f;
+    int samplingStep = max(1, (int)(1.f / volume->samplingRate));
+    int samplingCounter = 0;
+    float shadowIntensity = 0.f;
+
+    t0 += (0.2f * getRandomValue(sample, self->randomNumber)) - 0.1f;
+    t1 = min(ray.t, t1);
+
+    for (float t = t0; t < t1 && pathColor.w < 0.9f; t += epsilon)
     {
-        //        const vec3f point = ((ray.org + ray.dir * t) -
-        //        self->volumeOffset) /
-        //                            self->volumeElementSpacing;
+        vec3f volumeSampleColor;
+        float sampleOpacity = 0.f;
+
         const vec3f point = ray.org + t * ray.dir;
+        const float volumeSample = volume->sample(volume, point);
 
-        //        if (point.x >= 0.f && point.x < self->volumeDimensions.x &&
-        //            point.y >= 0.f && point.y < self->volumeDimensions.y &&
-        //            point.z >= 0.f && point.z < self->volumeDimensions.z)
-        {
-            const float sample = volume->sample(volume, point);
+        // Look up the opacity associated with the volume sample.
+        sampleOpacity = volume->transferFunction->getOpacityForValue(
+            volume->transferFunction, volumeSample);
+        if (samplingCounter == 0 && sampleOpacity < 0.01f)
```

Comment:
Make magic number 0.01f a constant somewhere?

Reply:
Also made it a renderer attribute (VolumeSamplingThreshold)